gnucash/gnome/dialog-stock-editor.c

```diff
@@ -0,0 +1,689 @@
+\********************************************************************/
+void gnc_ui_stockeditor_window_raise (StockEditorWindow * stockeditorData)
+{
+    if (stockeditorData && stockeditorData->window)
+        gtk_window_present (GTK_WINDOW (stockeditorData->window));
+}
+
+static void gnc_stockeditor_set_title_name (GtkWidget *window, Account *account)
+{
+    gchar *fullname = gnc_account_get_full_name (account);
+    gchar *title = g_strconcat(fullname, " - ", _("Stock Editor"), NULL);
+    gtk_window_set_title (GTK_WINDOW (window), title);
+    g_free (fullname);
+    g_free (title);
+}
+
+void cancel_button_cb (GtkWidget *widget, StockEditorWindow *data)
+{
+    gtk_widget_destroy (data->window);
+    g_free (data);
+}
+
+static gboolean gae_unfocus (GtkWidget *widget, const StockEditorWindow *data)
+{
+    gnc_amount_edit_evaluate (GNC_AMOUNT_EDIT (widget));
+    return FALSE;
+}
+
+
+static Account * gas_get_account (GtkWidget *gae)
+{
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return NULL;
+    return gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gae));
+}
+
+static gnc_numeric gae_get_amount (GNCAmountEdit *gae)
+{
+    gnc_numeric retval = gnc_numeric_zero ();
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae))) return retval;
+    if (!gnc_amount_edit_expr_is_valid (gae, &retval, FALSE)) return retval;
+    return gnc_numeric_zero ();
+}
+
+static gboolean req_acc_missing (GtkWidget *gae)
+{
+    if (!gtk_widget_get_sensitive (gae)) return FALSE;
+    return (gnc_account_sel_get_account (GNC_ACCOUNT_SEL (gae)) == NULL);
+}
+
+static gchar * append_status (gchar *prev_str, gchar *str)
+{
+    gchar *newstr;
+    if (!prev_str) return str;
+
+    newstr = g_strconcat (prev_str, "\n", str, NULL);
+    g_free (str);
+    g_free (prev_str);
+    return newstr;
+}
+
+static void check_signs (GNCAmountEdit *gae, gchar mask, gchar **status,
+                         gboolean *passes, gchar *type)
+{
+    gchar *sign;
+    gnc_numeric num;
+
+    /* mask=7 is permissive. no need to check anything. */
+    if (!gtk_widget_get_sensitive (GTK_WIDGET (gae)) || (mask == '7'))
+        return;
+
+    num = gae_get_amount (gae);
+    if (gnc_numeric_positive_p (num) &&
+        (mask == '1' || mask == '3' || mask == '#' || mask == '5'))
+        return;
+    if ((gnc_numeric_zero_p (num)) &&
+        (mask == '2' || mask == '3' || mask == '#' || mask == '6'))
+        return;
+    if (gnc_numeric_negative_p (num) &&
+        (mask == '4' || mask == '5' || mask == '6'))
+        return;
+
+    if (mask == '1') sign = "positive";
+    else if (mask == '2') sign = "zero";
+    else if (mask == '3') sign = "zero or positive";
+    else if (mask == '#') sign = "zero or positive";
+    else if (mask == '4') sign = "negative";
+    else if (mask == '5') sign = "negative or positive";
+    else if (mask == '6') sign = "zero or negative";
+    *status = append_status (*status, g_strdup_printf ("%s must be %s", type, sign));
```

Comment:
Does this need translating?

Reply:
No, it needs redesigning. It's about as wrong as it's possible to be.

Reply:
@jralls would you mind writing a template bitwise mask in `initialize_actions` and I'll convert other actions? e.g. a regular BUY.

Reply:
@christopherlam no. The problem with this PR isn't that you don't know how to do bit masks, the problem is that you've decomposed the problem the wrong way. Giving you a plug-and-chug bit mask solution will neither fix the wrong decomposition nor teach you basic CS.